pkg/cmd/util/clientcmd/factory.go

```diff
@@ -367,57 +351,54 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
 	}
 	kLogsForObjectFunc := w.Factory.LogsForObject
 	w.LogsForObject = func(object, options runtime.Object) (*restclient.Request, error) {
-		switch t := object.(type) {
-		case *deployapi.DeploymentConfig:
-			dopts, ok := options.(*deployapi.DeploymentLogOptions)
-			if !ok {
-				return nil, errors.New("provided options object is not a DeploymentLogOptions")
-			}
-			oc, _, err := w.Clients()
-			if err != nil {
-				return nil, err
-			}
-			return oc.DeploymentLogs(t.Namespace).Get(t.Name, *dopts), nil
-		case *buildapi.Build:
-			bopts, ok := options.(*buildapi.BuildLogOptions)
-			if !ok {
-				return nil, errors.New("provided options object is not a BuildLogOptions")
-			}
-			if bopts.Version != nil {
-				return nil, errors.New("cannot specify a version and a build")
-			}
-			oc, _, err := w.Clients()
-			if err != nil {
-				return nil, err
-			}
-			return oc.BuildLogs(t.Namespace).Get(t.Name, *bopts), nil
-		case *buildapi.BuildConfig:
-			bopts, ok := options.(*buildapi.BuildLogOptions)
-			if !ok {
-				return nil, errors.New("provided options object is not a BuildLogOptions")
-			}
+		gvk, _, err := api.Scheme.ObjectKind(object)
+		if err != nil {
+			return nil, err
+		}
+		if latest.OriginKind(gvk) {
 			oc, _, err := w.Clients()
 			if err != nil {
 				return nil, err
 			}
-			builds, err := oc.Builds(t.Namespace).List(api.ListOptions{})
-			if err != nil {
-				return nil, err
-			}
-			builds.Items = buildapi.FilterBuilds(builds.Items, buildapi.ByBuildConfigPredicate(t.Name))
-			if len(builds.Items) == 0 {
-				return nil, fmt.Errorf("no builds found for %q", t.Name)
-			}
-			if bopts.Version != nil {
-				// If a version has been specified, try to get the logs from that build.
-				desired := buildutil.BuildNameForConfigVersion(t.Name, int(*bopts.Version))
-				return oc.BuildLogs(t.Namespace).Get(desired, *bopts), nil
+			switch t := object.(type) {
```

Comment:
If you switch on type you don't need the GVK check.   Copying the client load into each block is probably better (we don't need to load the client at all if the type isn't recognized.).